### PR TITLE
derp: rate limit how often same-key clients can kick each other off server

### PR DESCRIPTION
**Pull request by bradfitz**

### CODE REVIEW

File: derp/derp_server.go [hunk 1]
```diff
@@ -472,7 +495,15 @@ func (s *Server) accept(nc Conn, brw *bufio.ReadWriter, remoteAddr string, connN
...
 	}
 
-	s.registerClient(c)
+	for {
+		ok, d := s.registerClient(c)
+		if ok {
+			break
+		}
+		s.clientsReplaceSleeping.Add(1)
+		time.Sleep(d)
```
Comment (sailorfrag) [inline derp/derp_server.go:504]: should it sleep d + (something small) or is the rate limiter immune to scheduler quantization type issues where it wakes up only to discover it needs to sleep a bit longer?

Comment (bradfitz) [reply derp/derp_server.go:504]: I don't think anything else is needed. Never seen other x/time/rate users do that. And if there's a thundering herd, it's of a relatively small number of clients per herd. (and usually it's just one, as two nodes fighting is most common)

**Review by crawshaw — APPROVED** [review]
I think that will work! Going to stick with manual testing?

**Comment by bradfitz** [comment]
I added tests which found bugs and existing problems, yay. Converting to a draft for a bit.

**Pull request merged**